src/prefect/cli/cloud.py

```diff
@@ -35,110 +50,244 @@
-        f"/accounts/{workspace['account_id']}"
-        f"/workspaces/{workspace['workspace_id']}"
-    )
+def set_login_api_ready_event():
+    login_api.extra["ready-event"].set()
+
+
+login_api = FastAPI(on_startup=[set_login_api_ready_event])
+"""
+This small API server is used for data transmission for browser-based log in.
+"""
+
+login_api.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+
+class LoginSuccess(BaseModel):
+    api_key: str
+
+
+class LoginFailed(BaseModel):
+    reason: str
+
+
+class LoginResult(BaseModel):
+    type: Literal["success", "failure"]
+    content: Union[LoginSuccess, LoginFailed]
+
+
+class ServerExit(Exception):
+    pass
+
+
+@login_api.post("/success")
+def receive_login(payload: LoginSuccess):
+    login_api.extra["result"] = LoginResult(type="success", content=payload)
+    login_api.extra["result-event"].set()
+
+
+@login_api.post("/failure")
+def receive_failure(payload: LoginFailed):
+    login_api.extra["result"] = LoginResult(type="failure", content=payload)
+    login_api.extra["result-event"].set()
+
+
+async def serve_login_api(cancel_scope, task_status):
+    config = uvicorn.Config(login_api, port=0, log_level="critical")
+    server = uvicorn.Server(config)
+
+    try:
+        # Yield the server object
+        task_status.started(server)
+        await server.serve()
+    except anyio.get_cancelled_exc_class():
+        pass  # Already cancelled, do not cancel again
+    except SystemExit as exc:
+        # If uvicorn is misconfigured, it will throw a system exit and hide the exc
+        app.console.print("[red][bold]X Error starting login service!")
+        cause = exc.__context__  # Hide the system exit
+        traceback.print_exception(type(cause), value=cause, tb=cause.__traceback__)
+        cancel_scope.cancel()
+    else:
+        # Exit if we are done serving the API
+        # Uvicorn overrides signal handlers so without this Ctrl-C is broken
+        cancel_scope.cancel()
 
 
 def confirm_logged_in():
     if not PREFECT_API_KEY:
         profile = prefect.context.get_settings_context().profile
         exit_with_error(
             f"Currently not authenticated in profile {profile.name!r}. "
-            "Please login with `prefect cloud login --key <API_KEY>`."
+            "Please log in with `prefect cloud login`."
         )
 
 
-def get_current_workspace(workspaces):
-    workspace_handles_by_id = {
-        workspace[
-            "workspace_id"
-        ]: f"{workspace['account_handle']}/{workspace['workspace_handle']}"
-        for workspace in workspaces
-    }
-    current_workspace_id = re.match(
-        r".*accounts/.{36}/workspaces/(.{36})\Z", PREFECT_API_URL.value()
-    ).groups()[0]
-    return workspace_handles_by_id[current_workspace_id]
-
+def get_current_workspace(workspaces: Iterable[Workspace]) -> Optional[Workspace]:
+    current_api_url = PREFECT_API_URL.value()
```

Comment:
Is the `PREFECT_API_URL` always going to be the workspace? I realize that it could be pointed to a self-hosted orion instance, but do we treat those as their own self-contained workspaces?

More specifically, is there a need for us to distinguish between the API URL and the workspace?